src/EffectNode.cpp

```diff
@@ -15,105 +15,91 @@
 #include <algorithm>
 #include "Paths.h"
 
-EffectNode::EffectNode(Context *c, QString file)
-    : VideoNode(c)
-    , m_intensity(0)
-    , m_openGLWorker(new EffectNodeOpenGLWorker(this), &QObject::deleteLater)
-    , m_ready(false) {
+EffectNode::EffectNode(Context *context, QString file)
+    : VideoNode(new EffectNodePrivate(context))
+{
+    qRegisterMetaType<Chain>("Chain");
 
+    d()->m_openGLWorker = QSharedPointer<EffectNodeOpenGLWorker>(new EffectNodeOpenGLWorker(*this), &QObject::deleteLater);
     setInputCount(1);
     setFrequency(0);
-    m_periodic.setInterval(10);
-    m_periodic.start();
-    connect(&m_periodic, &QTimer::timeout, this, &EffectNode::periodic);
-
-    m_beatLast = context()->timebase()->beat();
-    m_realTimeLast = context()->timebase()->wallTime();
-    connect(m_openGLWorker.data(), &EffectNodeOpenGLWorker::initialized, this, &EffectNode::onInitialized);
+    d()->m_periodic.setInterval(10);
+    d()->m_periodic.start();
+    connect(&d()->m_periodic, &QTimer::timeout, this, &EffectNode::onPeriodic);
 
+    d()->m_beatLast = context->timebase()->beat();
+    d()->m_realTimeLast = context->timebase()->wallTime();
     if (!file.isEmpty()) setFile(file);
 }
 
 EffectNode::EffectNode(const EffectNode &other)
     : VideoNode(other)
-//    , m_programs(other.m_programs)
-    , m_intensity(other.m_intensity)
-    , m_intensityIntegral(other.m_intensityIntegral)
-    , m_beatLast(other.m_beatLast)
-    , m_realTime(other.m_realTime)
-    , m_realTimeLast(other.m_realTimeLast)
-    , m_openGLWorker(other.m_openGLWorker)
-    , m_ready(other.m_ready)
-    , m_frequency(other.m_frequency) {
-/*
-    auto k = other.m_renderStates.keys();
-    for (int i=0; i<k.count(); i++) {
-        auto otherRenderState = other.m_renderStates.value(k.at(i));
-        m_renderStates.insert(k.at(i), QSharedPointer<EffectNodeRenderState>(new EffectNodeRenderState(*otherRenderState)));
-   }*/
+{
+}
+
+EffectNode::EffectNode(QSharedPointer<EffectNodePrivate> other_ptr)
+    : VideoNode(other_ptr.staticCast<VideoNodePrivate>())
+{
 }
 
-EffectNode::~EffectNode() {
+EffectNode *EffectNode::clone() const {
+    return new EffectNode(*this);
+}
+
+QSharedPointer<EffectNodePrivate> EffectNode::d() const {
+    return d_ptr.staticCast<EffectNodePrivate>();
 }
 
 QJsonObject EffectNode::serialize() {
     QJsonObject o = VideoNode::serialize();
-    o.insert("file", m_file);
-    o.insert("intensity", m_intensity);
+    o.insert("file", d()->m_file);
+    o.insert("intensity", d()->m_intensity);
     return o;
 }
 
-void EffectNode::onInitialized() {
+void EffectNode::chainsEdited(QList<Chain> added, QList<Chain> removed) {
+    for (auto chain : added) {
+        auto result = QMetaObject::invokeMethod(d()->m_openGLWorker.data(), "addNewState", Q_ARG(Chain, chain));
```

Comment:
I don't see why we have to go through the MetaObject here: isn't `m_openGLWorker.data()` known to be a `EffectNodeOpenGLWorker`? Can't you just call `m_openGLWorker->addNewState(chain)`?

Reply:
oh I see, it's to do message passing across threads. I almost wish that could be more explicit but this is fine.

Reply:
yeah, I only use `invokeMethod` to traverse threads. Polymorphism should work fine for calling things in the same thread.